src/java.base/share/classes/jdk/internal/misc/CDS.java

```diff
@@ -195,4 +201,79 @@ private static void validateInputLines(String[] lines) {
+              testStr.contains("-XX:+RecordDynamicDumpInfo");
+    }
+
+    /**
+    * called from jcmd VM.cds to dump static or dynamic shared archive
+    * @param isStatic indicates dump static archive of dynnamic archive.
+    * @param fileName user input archive name, can be null.
+    */
+    private static void dumpSharedArchive(boolean isStatic, String fileName) throws Exception {
+        boolean DEBUG =  System.getProperty("CDS.Debug") == "true";
+        String archiveFile =  fileName != null ? fileName :
+            "java_pid" + ProcessHandle.current().pid() + (isStatic ? "_static.jsa" : "_dynamic.jsa");
+        if (DEBUG) {
+            System.out.println((isStatic ? "Static" : " Dynamic") + " dump to file " + archiveFile);
+        }
+        if (isStatic) {
+            String listFile = archiveFile + ".classlist";
+            dumpClassList(listFile);
+            String jdkHome = System.getProperty("java.home");
+            ArrayList<String> cmds = new ArrayList<String>();
+            cmds.add(jdkHome + File.separator + "bin" + File.separator + "java"); // java
+            cmds.add("-Xlog:cds");
+            cmds.add("-Xshare:dump");
+            cmds.add("-XX:SharedClassListFile=" + listFile);
+            cmds.add("-XX:SharedArchiveFile=" + archiveFile);
+            // All args in command line
+            String[] vmArgs = getVMArguments();
+            if (vmArgs != null) {
+                for (String arg : vmArgs) {
+                    if (arg != null && !containsExcludedFlags(arg)) {
+                        cmds.add(arg);
+                    }
+                }
+            }
+
+            if (DEBUG) {
+                System.out.println("Static dump cmd: ");
+                for (String s : cmds) {
+                    System.out.print(s + " ");
+                }
+                System.out.println("");
+            }
+
+            // Do not take parent env which will cause dumping fail.
+            Process proc = Runtime.getRuntime().exec(cmds.toArray(new String[0]),
+                               new String[] {"EnvP=null"});
+            if (DEBUG) {
+                System.out.println("Dumping process " + proc.pid() + " Stdout: ");
+                String line;
+                InputStreamReader isr = new InputStreamReader(proc.getInputStream());
+                BufferedReader rdr = new BufferedReader(isr);
+                while((line = rdr.readLine()) != null) {
+                    System.out.println(line);
+                }
+
+                System.out.println("Dumping process " + proc.pid() + " Stderr: ");
+                isr = new InputStreamReader(proc.getErrorStream());
+                rdr = new BufferedReader(isr);
+                while((line = rdr.readLine()) != null) {
+                    System.out.println(line);
+                }
+            }
+            proc.waitFor();
+        } else {
+            dumpDynamicArchive(archiveFile);
+        }
+    }
```

Comment:
I think we should have some error checks and clean up:

- Remove the classlist file
- Check if if the process exit status is 0
- Remove the JSA file first, then try to dump it, and check if the file exists afterwards. If not, report the error. (For both dynamic and static dumps)

Reply:
The classlist file is not deleted after the dump has finished.